Add archive page for older LTS changelogs and update recent changelog page

biru-codeastromer opened the pull request:
Fixes #7502

The changelog-stable page for LTS changelogs displayed all changelogs in a single list, without any filtering or separation for older entries. This caused a cluttered user experience, especially when trying to view recent updates. To address this, a feature was requested to separate recent and older changelogs.

### Changes Implemented
1. Filtering Changelogs:

   a. Updated index.html.haml to display only recent changelogs (from the past 2 years).
   b. Added filtering logic using Ruby to select recent entries (release.date > 2.years.ago).

2. Created Archive Page:

a)Added a new archive.html.haml file under content/changelog-stable/ to display older changelogs (older than 2 years).
b)Implemented filtering logic to show only archived entries (release.date <= 2.years.ago).

### Impacts
1. The main changelog page now displays only recent updates, reducing clutter and improving readability.
2. Older changelogs are moved to a dedicated archive page, ensuring that all changelogs remain accessible.

Hope this solves the issue @timja

@kmartens27 @gounthar Hello Sir's, could you please review these changes and let me know if everything looks good? Also, if there are any further improvements or adjustments I should consider, I'd appreciate your suggestions!..thanks

gounthar reviewed (changes requested):
Thanks for your contribution.
The CI fails, though, with two errors, one of them being:

> ```
>  An error during rendering /changelog-stable/archive.html.haml occurred.
> While processing file /home/jenkins/agent/workspace/Websites_jenkins.io_PR-7746/content/changelog-stable/archive.html.haml
> An error occurred: undefined method `years' for an instance of Integer
> ```

krisstern commented:
Hi @biru-codeastromer please test locally first to make sure the build can be generated successfully. Now I am seeing the below in the logs:

```console
[2024-12-11T08:25:59.332Z] Please see .awestruct/error.log for more information
[2024-12-11T08:25:59.332Z] An error during rendering /changelog-stable/index.html.haml occurred.
[2024-12-11T08:25:59.332Z] While processing file /home/jenkins/agent/workspace/Websites_jenkins.io_PR-7746/content/changelog-stable/index.html.haml
[2024-12-11T08:25:59.332Z] An error occurred: undefined method `years' for an instance of Integer
```

biru-codeastromer commented:
@kmartens27  @gounthar Hi Sir, I'm seeing the following errors in the CI checks:

1.PR merge build failed
2.Website job failed to build
3.Shell step error during site build


Could you please review the logs and help me identify the cause of these issues? Any guidance or suggestions for fixing them would be greatly appreciated!

Also i want to know if this is happening from my end or it is issue with the build system itself!

biru-codeastromer commented:
@krisstern Okay Sir looking into it now,thanks!

biru-codeastromer commented:
@kmartens27 sure Sir,I am working on it

biru-codeastromer commented:
Okay, Sir will  try to do your suggestion @MarkEWaite !

Also , @kmartens27 @krisstern @gounthar . Thank you for the detailed guidance earlier. I wanted to give an update and ask for some clarification regarding an issue I'm facing.

I’ve been working on fixing the filtering logic for recent and archived LTS changelogs based on release.date in lts.yml. The logic works correctly in terms of the code changes I’ve made, but I’m encountering an issue when running make generate locally on my machine. The release.date parsing seems to be causing errors or unexpected behavior during the generation step.

I’m trying to debug and correct the issue but wanted to check if there are any specific constraints or known issues with release.date formatting in lts.yml that I might have missed. Also, could there be local environment dependencies affecting how make generate processes these dates?

Any pointers or guidance would be much appreciated. Thanks in advance for your help!

MarkEWaite commented:
> I’ve been working on fixing the filtering logic for recent and archived LTS changelogs based on release.date in lts.yml.

The weekly changelog archive (changelog/ and changelog-old/) avoid date computations by using a count of releases instead of a date.  If you use a count of 24 or 25 releases, that is approximately two years of LTS releases (a release every 4 weeks with a two week break at the end of each calendar year).

biru-codeastromer commented:
thank you Sir @MarkEWaite for letting me know that switching to a release count aligns with the expectations for the LTS changelog. now will update the codes accordingly

biru-codeastromer commented:
Hi @MarkEWaite Sir,

I’ve completed the changes as discussed and raised the PR. The implementation includes using the same technique for the LTS changelog as for the weekly changelog, with a new changelog-stable-old directory created to archive older LTS releases. Additionally, the logic for generating the pages has been implemented in the respective index.html.haml files, and I’ve ensured proper navigation between changelog-stable and changelog-stable-old.

Also @kmartens27  @gounthar Sir,All checks have now passed successfully. Could you please review the PR and let me know if there are any further changes or improvements needed? Looking forward to your feedback.

Thank you!

biru-codeastromer commented:
Hi @MarkEWaite  Sir,

Thank you for pointing out the issue! I’ve successfully applied the necessary corrections to the changelog files. The changelog-stable now displays the most recent changelogs, and the changelog-stable-old handles the archived ones. 

I’ve also updated the app-banner links  to reflect these changes. Also added the app-banner links in changelog-stable-old to get back to the recent one.

The recent changelogs now use drop(25) and the archived ones use take(25) for proper segmentation, previously it was the other way  around.

Please review the latest changes at your convenience and let me know if there’s anything else to address. Looking forward to your feedback!

Thank you for your guidance!

zbynek reviewed (changes requested):
@biru-codeastromer 
> The recent changelogs now use drop(25) and the archived ones use take(25) 

That means the "old" changelog will always have 25 releases and the current one will keep growing. It would make sense to keep the number of items in the stable changelog fixed instead and let the "old" changelog grow. Also

```
= partial('changelog-stable.html')
```
is responsible for including the really old versions, but it's part of the stable changelog rather than old, meaning that "old" changelog now contains the middle 25 releases and stable has everything else.

biru-codeastromer commented:
Hi @zbynek ,

Thank you for the feedback! I’ve made the necessary changes as per your suggestion. The "stable" changelog now has a fixed number of items, while the "old" changelog will continue to grow with additional releases.I have reversed it .

I’ve also corrected the usage of partial('changelog-stable.html') by moving it to the old changelog, so that it now correctly includes the older versions within the stable changelog rather than the "old" changelog.

Please take a look at the updated changes and let me know if there’s anything else that needs adjustment.

Thanks again for your help!

biru-codeastromer commented:
@zbynek Hi Sir, I again changed the changelog-stable and changelog-stable-old since i saw that after deployment it is just becoming reverse.
I moved the  '= partial('changelog-stable.html')' to changelog-stable-old.

zbynek commented:
> I again changed the changelog-stable and changelog-stable-old since i saw that after deployment it is just becoming reverse.

By combining `reverse`, `drop` and `take` properly you can achieve a state where one page contains the latest 25 releases (in descending order) and the other one contains all the remining ones. I'd suggest to take your time to set up a local build and check your changes before pushing.

biru-codeastromer commented:
thank you Sir's @krisstern  @MarkEWaite  for guidance, will now work to sort it out!


thank you for your help Sir @zbynek I will take my time and check throughoughly in local before pushing again

biru-codeastromer commented:
Hi @krisstern, @MarkEWaite, @zbynek,

I have made the necessary changes to the changelog pages. The changelog-stable now correctly shows only the releases from the last two years (2024 and 2023), and the changelog-stable-old contains all older releases starting from 2022 and before.Also added comments for better readability.

I’ve verified these changes locally, and they should now align with the intended functionality. Please review the changes and let me know if everything looks good or if further adjustments are required.

Thank you for your guidance!

krisstern commented on content/changelog-stable/index.html.haml:
```diff
@@ -11,22 +11,31 @@ actions:
 
-  = partial('changelog-stable.html')
+.app-banner
+  Changelogs of historical LTS releases can be found
```
```suggestion
  Changelogs of LTS releases of more than two years old can be found
```

krisstern commented on content/changelog-stable-old/index.html.haml:
```diff
@@ -0,0 +1,28 @@
+        %p No notable changes for this release.
+
+.app-banner
+  Recent LTS changelogs can be found
```
```suggestion
  More recent LTS changelogs of less than two years old can be found
```

krisstern reviewed (changes requested):
@biru-codeastromer I think we are almost achieving our goals, except for some edits requested.

biru-codeastromer commented:
@krisstern Hi Sir, glad that we are achieving what we needed.I have committed your changes now . Hope it fits in! Thanks

krisstern commented on content/changelog-stable-old/index.html.haml:
```diff
@@ -0,0 +1,28 @@
+---
+layout: changelog
+title: Archived LTS Changelogs
```
```suggestion
title: Archived LTS Changelog
```

krisstern commented on content/changelog-stable-old/index.html.haml:
```diff
@@ -0,0 +1,28 @@
+
+.ratings
+  - # Load older changelogs (all except the most recent 25, ordered correctly)
+  - archived_changelogs = site.changelogs[:lts].reverse.drop(27)
```
Is there a reason why the `27` is hard-coded?

krisstern replied:
And, why 27?

biru-codeastromer replied:
Hi Sir @krisstern  ,I observed that using 25 excluded the January 2023 releases from the correct page when I saw in local build, so I adjusted it to 27 to include them. I hard-coded 27 as a  fix to show latest 2 years list, @MarkEWaite Sir told me about this that instead of date computations ,release of counts are used. Let me know your thoughts if it is fine or can be improved !

krisstern replied:
I don't think that was exactly what Mark said. Direct quote:
>  The weekly changelog archive (changelog/ and changelog-old/) avoid date computations by using a count of releases instead of a date. If you use a count of 24 or 25 releases, that is approximately two years of LTS releases (a release every 4 weeks with a two week break at the end of each calendar year).

The issue is I don't know why we have the number 25 before, but then you seem to need 27...

krisstern replied:
I tend to think we should keep 25... unless the other maintainers have a different view

biru-codeastromer replied:
Hi @krisstern ,

Thank you for pointing this out. I understand now that the original intent was to approximate two years of LTS releases, which aligns with using 24 or 25 releases. My change to 27 was primarily to ensure that the January 2023 releases were included in the list during my local testing since then we could take exact two years starting from 2023.

Based on your feedback, I’ll revert this back to 25 for consistency Or I should wait for other maintainer view too Sir?

Let me know if you have any additional thoughts or suggestions!

krisstern commented on content/changelog-stable/index.html.haml:
```diff
@@ -11,22 +11,31 @@ actions:
-  - # source: https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/lts.yml
-  - site.changelogs[:lts].reverse_each do | release |
+  - # Load the most recent 25 changelogs (latest first)
+  - recent_changelogs = site.changelogs[:lts].reverse.take(27)
```
Same as above: Is there a reason why the `27` is hard-coded?

kmartens27 replied:
I agree with @krisstern that 25 is more than enough in regards to what should be shown. The oldest Jenkins LTS version that is typically encouraged/supported would be included in this without having too many additional versions displayed.

biru-codeastromer commented:
I have committed all your suggestions @krisstern !Thanks for that Sir

krisstern commented on content/changelog-stable-old/index.html.haml:
```diff
@@ -0,0 +1,28 @@
+
+.ratings
+  - # Load older changelogs (all except the most recent 25, ordered correctly)
+  - archived_changelogs = site.changelogs[:lts].reverse.drop(27)
```
```suggestion
  - archived_changelogs = site.changelogs[:lts].reverse.drop(25)
```

krisstern commented on content/changelog-stable/index.html.haml:
```diff
@@ -11,22 +11,31 @@ actions:
-  - # source: https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/lts.yml
-  - site.changelogs[:lts].reverse_each do | release |
+  - # Load the most recent 25 changelogs (latest first)
+  - recent_changelogs = site.changelogs[:lts].reverse.take(27)
```
```suggestion
  - recent_changelogs = site.changelogs[:lts].reverse.take(25)
```

krisstern reviewed (changes requested):
We are almost there

biru-codeastromer commented:
okay Sir 's @krisstern @kmartens27

 I have committed your suggestion Sir and now it changed to 25 releases ! Hope it sorts it out.

zbynek commented on content/changelog-stable-old/index.html.haml:
```diff
@@ -0,0 +1,28 @@
+        %p No notable changes for this release.
+
+.app-banner
+  More recent LTS changelog of less than two years old can be found
```
If the split between changelog pages is defined by a number of releases and not by time, it would be better to simplify this.

```suggestion
  This is the changelog archive. Changelog for recent releases can be found
```

zbynek commented on content/changelog-stable/index.html.haml:
```diff
@@ -11,22 +11,31 @@ actions:
-          = partial('changelog-changes.html.haml', :changes => release.lts_changes)
+          = partial('../_partials/changelog-changes.html.haml', :changes => release.lts_changes)
+
+%a.archive-link{:href => '/changelog-stable-old/'}
```
Duplicates the banner below => remove.

biru-codeastromer replied:
I have removed it Sir  @zbynek

zbynek commented on content/changelog-stable/index.html.haml:
```diff
@@ -11,22 +11,31 @@ actions:
 
-  = partial('changelog-stable.html')
+.app-banner
+  Changelog of LTS releases of more than two years old can be found
```
```suggestion
  Changelogs of historical releases can be found
```

kmartens27 commented on content/changelog-stable-old/index.html.haml:
```diff
@@ -0,0 +1,28 @@
+        %p No notable changes for this release.
+
+.app-banner
+  This is the changelog archive. Changelog for recent releases can be found
```
```suggestion
  This is the changelog archive. Changelogs for recent releases can be found
```
Adjust plural usage as there are multiple changelogs.

zbynek replied:
This is the very opposite of what @krisstern suggested in their previous review and I tend to agree:
* changelog = document containing info about multiple releases
* changelog entry = part of changelog corresponding to a single release

kmartens27 replied:
```suggestion
  This is the changelog archive. The changelog for recent releases can be found
```
Ah sorry about that. In that case it should have something in front of it like this.

biru-codeastromer replied:
I have updated every suggestion Sir @kmartens27

kmartens27 commented on content/changelog-stable/index.html.haml:
```diff
@@ -11,22 +11,31 @@ actions:
 
-  = partial('changelog-stable.html')
+.app-banner
+  Changelog of historical releases can be found
```
```suggestion
  Changelogs of historical releases can be found
```
Same as above, adjusting for plural usage.

krisstern commented:
If we are going to be consistent, might as well change the titles of the two files as well to use the plural instead of the singular?

biru-codeastromer commented:
> If we are going to be consistent, might as well change the titles of the two files as well to use the plural instead of the singular?

Hi Sir @krisstern I have changed the files name and also updated the links in the index.html.haml files to match the file names .

zbynek commented:
Please don't change the filenames -- `/changelog` and `/changelog-stable` are both linked from many places on the website as well as from Jenkins itself and potentially many external websites.

biru-codeastromer commented:
> Please don't change the filenames -- `/changelog` and `/changelog-stable` are both linked from many places on the website as well as from Jenkins itself and potentially many external websites.

@zbynek Done it Sir

biru-codeastromer commented:
@krisstern I have done every suggestions Sir . Hope it is good now !
If any more refining needed , I am ready for suggestions. Thanks .

krisstern commented:
Just don't change filenames unless you know for sure it is okay to do so... The filenames normally follow a particular pattern so if you change them you would be breaking something, like what you just witnessed.

biru-codeastromer commented:
> Just don't change filenames unless you know for sure it is okay to do so... The filenames normally follow a particular pattern so if you change them you would be breaking something, like what you just witnessed.

Hi Sir @krisstern i have back it again and fix it !  Thanks for telling it , I will  always keep that in mind  .

krisstern reviewed (approved):
Thanks so much for the hard work @biru-codeastromer!

biru-codeastromer commented:
Hi Sir's @krisstern  @kmartens27  @MarkEWaite  @zbynek  @gounthar ,

I want to take a moment to thank all the maintainers for your constant support and guidance throughout my contributions. Your detailed reviews, suggestions, and timely responses have made this experience incredibly enriching and enjoyable.
Contributing to Jenkins has been a fantastic journey, and I’ve learned so much from all of you . It  inspires me to keep improving and contributing.

I am hopeful that this PR is now completed and completes the issue #7502  , and I’m hopeful it will be merged soon. Regardless, if more suggestions I am ready.

Looking forward to contributing more to Jenkins and learning from this amazing team!
Thank you again !

MarkEWaite reviewed (approved):
Looks great.  Thanks

biru-codeastromer commented:
> Looks great. Thanks

Welcome Sir @MarkEWaite !

biru-codeastromer commented:
Hi  @gounthar Sir ,
I hope you're doing well. I hope your required changes are met .Please let me know if there’s anything I can clarify or improve further!

Thanks !

biru-codeastromer commented:
<img width="1512" alt="Screenshot 2024-12-14 at 8 57 59 PM" src="https://github.com/user-attachments/assets/8a1c0430-1345-41cf-b0dd-384d81928828" />

<img width="1512" alt="Screenshot 2024-12-14 at 8 58 21 PM" src="https://github.com/user-attachments/assets/2b04e2f0-4803-46bd-abad-622c74dd5a82" />


Hi Sir @MarkEWaite @zbynek ,
I hope you're doing well. I wanted to check something regarding my pull request. I noticed that a "Request Changes" comment was added, referring to a file/section that I had already deleted or replaced in the latest commits.

Could you kindly help me understand why this is happening? Do I need to address these changes again, or can they be dismissed since the file no longer exists in the current state?
Thank you for your guidance and time! Please let me know if there's anything specific I need to do to proceed.

MarkEWaite commented:
> I hope you're doing well. I wanted to check something regarding my pull request. I noticed that a "Request Changes" comment was added, referring to a file/section that I had already deleted or replaced in the latest commits.
> 
> Could you kindly help me understand why this is happening? Do I need to address these changes again, or can they be dismissed since the file no longer exists in the current stat

That entry indicates that @gounthar had requested changes in one or more previous changes.  That request for changes is not cleared until @gounthar reviews it or one of the other maintainers dismisses the request for changes.  There is nothing that you need to do at this time.

biru-codeastromer commented:
> That entry indicates that @gounthar had requested changes in one or more previous changes. That request for changes is not cleared until @gounthar reviews it or one of the other maintainers dismisses the request for changes. There is nothing that you need to do at this time.

@MarkEWaite thank you Sir for the  confirmation .

Thank You Sir for the review too @gounthar !

kmartens27 commented:
@biru-codeastromer I'm not sure what the specific issue or code is that is preventing the full changelog from being displayed on the archive, but it would need to be corrected, otherwise we will need to revert the archive creation until it can be properly rendered. I also wanted to move the conversation to here where it would be more relevant. Sorry for jumping between several tickets.

biru-codeastromer commented:
Hi @kmartens27 , 
Thanks for moving the conversation here! I'll take a closer look at how `.1` releases are being processed and displayed. Since my PR mainly introduced the archive layout, the issue might be related to how `lts.yml` stores `.1` releases or how they're being pulled for rendering.

I’ll check if `.1` releases in `lts.yml` have complete `lts_changes` data, whether they are being correctly fetched in `site.changelogs[:lts]` , if the rendering logic needs adjustments.

I'll update soon after investigating! 🚀

biru-codeastromer commented:
But I don't think the addition of this PR has hampered that . What's your view on this ?

daniel-beck commented on content/changelog-stable/index.html.haml:
```diff
@@ -11,22 +11,27 @@ actions:
-          = partial('changelog-changes.html.haml', :changes => release.lts_changes)
+          = partial('../_partials/changelog-changes.html.haml', :changes => release.lts_changes)
 
-  = partial('changelog-stable.html')
```
Caused https://github.com/jenkins-infra/jenkins.io/issues/8040

biru-codeastromer replied:
thanks  a lot for the follow up @daniel-beck
